src/relax/op/tensor/transform.cc

```diff
@@ -0,0 +1,441 @@
+                          "single \"-1\". However, the given new shape is "
+                       << call->args[1]);
+    } else {
+      new_shape_prod = new_shape_prod * new_shape[i];
+    }
+  }
+  return {new_shape_prod, dim_to_infer};
+}
+
+StructInfo InferStructInfoReshape(const Call& call, const BlockBuilder& ctx) {
+  if (call->args.size() > 2) {
+    ctx->ReportFatal(Diagnostic::Error(call) << "Reshape op should no more than 2 arguments");
+  }
+  const auto* data_sinfo = GetStructInfoAs<TensorStructInfoNode>(call->args[0]);
+  if (data_sinfo == nullptr) {
+    ctx->ReportFatal(Diagnostic::Error(call)
+                     << "Reshape requires the input data to be Tensor. However, the given one is "
+                     << call->args[0]->struct_info_->GetTypeKey());
+  }
+
+  Optional<PrimExpr> old_shape_prod;
+  if (data_sinfo->shape.defined()) {
+    const auto* old_shape_sinfo = GetStructInfoAs<ShapeStructInfoNode>(data_sinfo->shape.value());
+    ICHECK_NOTNULL(old_shape_sinfo);
+    if (old_shape_sinfo->values.defined()) {
+      PrimExpr _old_shape_prod = IntImm(DataType::Int(64), 1);
+      for (int i = 0; i < data_sinfo->ndim; ++i) {
+        _old_shape_prod = _old_shape_prod * old_shape_sinfo->values.value()[i];
+      }
+      old_shape_prod = _old_shape_prod;
+    }
+  }
+
+  arith::Analyzer* analyzer = ctx->GetAnalyzer();
+  if (call->args.size() == 2) {
+    ICHECK(call->attrs.get() == nullptr);
+    const auto* new_shape_sinfo = GetStructInfoAs<ShapeStructInfoNode>(call->args[1]);
+    if (new_shape_sinfo == nullptr) {
+      ctx->ReportFatal(
+          Diagnostic::Error(call)
+          << "Reshape requires the input new shape to be Shape. However, the given one is "
+          << call->args[1]->struct_info_->GetTypeKey());
+    }
+
+    if (new_shape_sinfo->values.defined() && old_shape_prod.defined()) {
+      PrimExpr new_shape_prod = IntImm(DataType::Int(64), 1);
+      for (int i = 0; i < new_shape_sinfo->ndim; ++i) {
+        new_shape_prod = new_shape_prod * new_shape_sinfo->values.value()[i];
+      }
+      if (analyzer->CanProve(old_shape_prod.value() != new_shape_prod)) {
+        ctx->ReportFatal(Diagnostic::Error(call)
+                         << "Reshape expects the new shape to be convertible from the old shape. "
+                            "However, the old shape is "
+                         << data_sinfo->shape << ", with product " << old_shape_prod.value()
+                         << ", while the new shape is " << call->args[1] << ", with product "
+                         << new_shape_prod);
+      }
+    }
+    return TensorStructInfo(call->args[1], data_sinfo->dtype);
+  }
+
+  const auto* attrs = call->attrs.as<ReshapeAttrs>();
+  ICHECK_NOTNULL(attrs);
+  Array<PrimExpr> new_shape = attrs->new_shape;
+  if (!old_shape_prod.defined()) {
+    CheckNewShape(call, ctx, new_shape, /*allow_dim_inference=*/false);
```

Comment:
In this case, we should fallback to safety net, and return Tensor(ndim=new_shape.size())

Reply:
Add a regressiojn test

Reply:
Could you elaborate the reason behind more? My consideration is that it’s okay to trust the user input shape here, as long as it doesn’t contain any -1.

Reply:
My bad, I think it is OK to take new_shape